File: frontend/lib/calendar.ts
import { api } from "@/lib/api-client";

// ---------------------------------------------------------------- vocabularies

/** Mirrors CALENDAR_ENTRY_STATUSES in backend/app/models/calendar.py (order matters). */
export const CALENDAR_STATUSES = [
  "planned",
  "in_progress",
  "ready",
  "scheduled",
  "published",
  "cancelled",
] as const;
export type CalendarStatus = (typeof CALENDAR_STATUSES)[number];

export const CALENDAR_STATUS_LABELS_FA: Record<string, string> = {
  planned: "برنامه‌ریزی‌شده",
  in_progress: "در حال تولید",
  ready: "آماده بازبینی",
  scheduled: "زمان‌بندی‌شده",
  published: "منتشر شده",
  cancelled: "لغو شده",
};

/** Tailwind classes per status, for column headers and chips. */
export const CALENDAR_STATUS_STYLE: Record<string, string> = {
  planned: "bg-slate-500/15 text-slate-300 border-slate-500/30",
  in_progress: "bg-amber-500/15 text-amber-300 border-amber-500/30",
  ready: "bg-sky-500/15 text-sky-300 border-sky-500/30",
  scheduled: "bg-indigo-500/15 text-indigo-300 border-indigo-500/30",
  published: "bg-emerald-500/15 text-emerald-300 border-emerald-500/30",
  cancelled: "bg-red-500/15 text-red-300 border-red-500/30",
};

export const CALENDAR_PRIORITIES = ["low", "normal", "high", "urgent"] as const;
export type CalendarPriority = (typeof CALENDAR_PRIORITIES)[number];

export const CALENDAR_PRIORITY_LABELS_FA: Record<string, string> = {
  low: "کم",
  normal: "معمولی",
  high: "زیاد",
  urgent: "فوری",
};

export const CALENDAR_PRIORITY_STYLE: Record<string, string> = {
  low: "bg-white/5 text-muted-foreground",
  normal: "bg-sky-500/15 text-sky-300",
  high: "bg-amber-500/15 text-amber-300",
  urgent: "bg-red-500/15 text-red-300",
};

export const CALENDAR_SOURCE_LABELS_FA: Record<string, string> = {
  manual: "دستی",
  ai_auto: "زمان‌بندی AI",
};

// -------------------------------------------------------------------- entities

/** Mirrors CalendarEntryRead in backend/app/schemas/calendar.py. */
export interface CalendarEntry {
  id: string;
  organization_id: string;
  website_id: string;
  title: string;
  brief_id: string | null;
  article_id: string | null;
  opportunity_id: string | null;
  status: CalendarStatus | string;
  priority: CalendarPriority | string;
  source: string;
  scheduled_for: string | null;
  deadline: string | null;
  published_at: string | null;
  assigned_to: string | null;
  target_keyword: string | null;
  notes: string | null;
  details: Record<string, any>;
  created_at: string;
  updated_at: string;
}

/** Mirrors CalendarBoardView: columns keyed by status. */
export interface CalendarBoardView {
  columns: Record<string, CalendarEntry[]>;
}

export interface CalendarSummary {
  by_status: Record<string, number>;
  overdue: number;
  due_this_week: number;
  unassigned: number;
}

export interface CalendarAutoScheduleResult {
  website_id: string;
  created: number;
  skipped: number;
  /** Open opportunities still without a slot after this run. */
  remaining_open: number;
  scheduled_through: string | null;
}

export interface CalendarDayBucket {
  date: string;
  entries: CalendarEntry[];
  count: number;
}

export interface CalendarMonthView {
  website_id: string;
  year: number;
  month: number;
  range_start: string;
  range_end: string;
  days: CalendarDayBucket[];
}

export interface CalendarWeekView {
  website_id: string;
  range_start: string;
  range_end: string;
  days: CalendarDayBucket[];
}

// ----------------------------------------------------------------- write bodies

export interface CalendarEntryCreateBody {
  website_id: string;
  title: string;
  status?: CalendarStatus;
  priority?: CalendarPriority;
  scheduled_for?: string | null;
  deadline?: string | null;
  target_keyword?: string | null;
  notes?: string | null;
}

export interface CalendarEntryUpdateBody {
  title?: string;
  status?: CalendarStatus;
  priority?: CalendarPriority;
  scheduled_for?: string | null;
  deadline?: string | null;
  target_keyword?: string | null;
  notes?: string | null;
  assigned_to?: string | null;
}

// ----------------------------------------------------------------------- calls

export function getCalendarBoard(websiteId: string) {
  return api.get<CalendarBoardView>(`/calendar/board?website_id=${websiteId}`);
}

export function getCalendarMonth(websiteId: string, year: number, month: number) {
  return api.get<CalendarMonthView>(`/calendar/month?website_id=${websiteId}&year=${year}&month=${month}`);
}

export function getCalendarWeek(websiteId: string, startDate: string) {
  return api.get<CalendarWeekView>(`/calendar/week?website_id=${websiteId}&start_date=${startDate}`);
}

export function listCalendarEntries(websiteId: string, params: Record<string, any> = {}) {
  const query = new URLSearchParams({ website_id: websiteId });
  Object.entries(params).forEach(([k, v]) => {
    if (v !== undefined && v !== null && v !== "") {
      query.set(k, String(v));
    }
  });
  return api.get<{ data: CalendarEntry[]; meta: { total: number; limit: number; offset: number } }>(
    `/calendar?${query.toString()}`
  );
}

export function getCalendarEntry(entryId: string) {
  return api.get<CalendarEntry>(`/calendar/${entryId}`);
}

export function getCalendarSummary(websiteId: string) {
  return api.get<CalendarSummary>(`/calendar/summary?website_id=${websiteId}`);
}

export function createCalendarEntry(body: CalendarEntryCreateBody) {
  return api.post<CalendarEntry>(`/calendar`, body);
}

export function updateCalendarEntry(
  entryId: string,
  body: CalendarEntryUpdateBody
) {
  return api.patch<CalendarEntry>(`/calendar/${entryId}`, body);
}

/** Board drag: change status and/or scheduled date only. */
export function moveCalendarEntry(
  entryId: string,
  body: { status?: CalendarStatus; scheduled_for?: string | null }
) {
  return api.post<CalendarEntry>(`/calendar/${entryId}/move`, body);
}

export function deleteCalendarEntry(entryId: string) {
  return api.delete<{ deleted: boolean; id: string }>(`/calendar/${entryId}`);
}

export function autoScheduleFromOpportunities(
  websiteId: string,
  maxEntries = 10
) {
  return api.post<CalendarAutoScheduleResult>(
    `/calendar/auto-schedule?website_id=${websiteId}`,
    { max_entries: maxEntries }
  );
}

File: backend/app/schemas/calendar.py
"""Schemas for the content calendar.

Write models are narrow on purpose. `organization_id` always comes from the
authenticated membership (never the body) and `published_at` / `source` are
system-set: `published_at` is stamped by `move_entry` when a slot reaches
`published`, and `source` is `manual` for anything created through this API —
only `auto_schedule_from_opportunities` may write `ai_auto`. If a client could
set either directly, a human-created slot could impersonate an AI one (or
fake a publish timestamp that never happened).
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.calendar import (
    CALENDAR_ENTRY_PRIORITIES,
    CALENDAR_ENTRY_SOURCES,
    CALENDAR_ENTRY_STATUSES,
)


# -------------------------------------------------------------------- read model

class CalendarEntryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    website_id: UUID
    title: str
    brief_id: UUID | None = None
    article_id: UUID | None = None
    opportunity_id: UUID | None = None
    status: str
    priority: str
    source: str
    scheduled_for: datetime | None = None
    deadline: datetime | None = None
    published_at: datetime | None = None
    assigned_to: UUID | None = None
    target_keyword: str | None = None
    notes: str | None = None
    details: dict = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


# ------------------------------------------------------------------- write models

class CalendarEntryCreate(BaseModel):
    """Plan a new slot. `website_id` comes from the query/body, not inferred."""

    website_id: UUID
    title: str = Field(min_length=3, max_length=500)
    brief_id: UUID | None = None
    article_id: UUID | None = None
    opportunity_id: UUID | None = None
    status: str = "planned"
    priority: str = "normal"
    scheduled_for: datetime | None = None
    deadline: datetime | None = None
    assigned_to: UUID | None = None
    target_keyword: str | None = Field(default=None, max_length=255)
    notes: str | None = Field(default=None, max_length=5000)
    details: dict = Field(default_factory=dict)

    @field_validator("status")
    @classmethod
    def _known_status(cls, v: str) -> str:
        if v not in CALENDAR_ENTRY_STATUSES:
            raise ValueError(f"status must be one of {sorted(CALENDAR_ENTRY_STATUSES)}")
        return v

    @field_validator("priority")
    @classmethod
    def _known_priority(cls, v: str) -> str:
        if v not in CALENDAR_ENTRY_PRIORITIES:
            raise ValueError(f"priority must be one of {sorted(CALENDAR_ENTRY_PRIORITIES)}")
        return v


class CalendarEntryUpdate(BaseModel):
    """Edit an existing slot. Every field optional so a PATCH can be partial."""

    title: str | None = Field(default=None, min_length=3, max_length=500)
    brief_id: UUID | None = None
    article_id: UUID | None = None
    opportunity_id: UUID | None = None
    status: str | None = None
    priority: str | None = None
    scheduled_for: datetime | None = None
    deadline: datetime | None = None
    assigned_to: UUID | None = None
    target_keyword: str | None = Field(default=None, max_length=255)
    notes: str | None = Field(default=None, max_length=5000)
    details: dict | None = None

    @field_validator("status")
    @classmethod
    def _known_status(cls, v: str | None) -> str | None:
        if v is not None and v not in CALENDAR_ENTRY_STATUSES:
            raise ValueError(f"status must be one of {sorted(CALENDAR_ENTRY_STATUSES)}")
        return v

    @field_validator("priority")
    @classmethod
    def _known_priority(cls, v: str | None) -> str | None:
        if v is not None and v not in CALENDAR_ENTRY_PRIORITIES:
            raise ValueError(f"priority must be one of {sorted(CALENDAR_ENTRY_PRIORITIES)}")
        return v


class CalendarEntryMove(BaseModel):
    """What a drag on the board or calendar grid does: change date and/or status."""

    scheduled_for: datetime | None = None
    status: str | None = None

    @field_validator("status")
    @classmethod
    def _known_status(cls, v: str | None) -> str | None:
        if v is not None and v not in CALENDAR_ENTRY_STATUSES:
            raise ValueError(f"status must be one of {sorted(CALENDAR_ENTRY_STATUSES)}")
        return v


# ---------------------------------------------------------------- aggregate views

class CalendarDayBucket(BaseModel):
    """One calendar day and the entries scheduled on it."""

    date: str  # ISO date string, e.g. "2026-08-14"
    entries: list[CalendarEntryRead] = Field(default_factory=list)
    count: int = 0


class CalendarMonthView(BaseModel):
    website_id: UUID
    year: int
    month: int
    range_start: datetime
    range_end: datetime
    days: list[CalendarDayBucket] = Field(default_factory=list)


class CalendarWeekView(BaseModel):
    website_id: UUID
    range_start: datetime
    range_end: datetime
    days: list[CalendarDayBucket] = Field(default_factory=list)


class CalendarBoardView(BaseModel):
    """Kanban columns, keyed by status in CALENDAR_ENTRY_STATUSES order."""

    columns: dict[str, list[CalendarEntryRead]] = Field(default_factory=dict)


class CalendarAutoScheduleRequest(BaseModel):
    """Trigger the AI scheduler for one website."""

    start_from: datetime | None = None
    max_entries: int = Field(default=10, ge=1, le=100)


class CalendarAutoScheduleResult(BaseModel):
    website_id: UUID
    created: int
    skipped: int
    # Open opportunities still without a slot after this run (they stay for the
    # next click - the scheduler adds at most max_entries per call).
    remaining_open: int = 0
    # True when the website already holds >= max_entries outstanding AI slots,
    # so this run deliberately created nothing instead of flooding the calendar.
    throttled: bool = False
    scheduled_through: datetime | None = None


class CalendarSummary(BaseModel):
    by_status: dict[str, int] = Field(default_factory=dict)
    overdue: int = 0
    due_this_week: int = 0
    unassigned: int = 0


__all__ = [
    "CalendarEntryRead",
    "CalendarEntryCreate",
    "CalendarEntryUpdate",
    "CalendarEntryMove",
    "CalendarDayBucket",
    "CalendarMonthView",
    "CalendarWeekView",
    "CalendarBoardView",
    "CalendarAutoScheduleRequest",
    "CalendarAutoScheduleResult",
    "CalendarSummary",
    "CALENDAR_ENTRY_STATUSES",
    "CALENDAR_ENTRY_PRIORITIES",
    "CALENDAR_ENTRY_SOURCES",
]
